src/core/ext/filters/client_channel/http_proxy.c

```diff
@@ -83,35 +84,47 @@ static bool proxy_mapper_map_name(grpc_exec_ctx* exec_ctx,
   }
   char* no_proxy_str = gpr_getenv("no_proxy");
   if (no_proxy_str != NULL) {
-    char *server_host;
-    char *server_port;
-    if (!gpr_split_host_port(uri->path[0] == '/' ? uri->path + 1 : uri->path, &server_host, &server_port)) {
-      gpr_log(GPR_INFO, "unable to split host and port, not checking no_proxy list for host '%s'", server_uri);
+    static const char* NO_PROXY_SEPARATOR = ",";
+    bool use_proxy = true;
+    char* server_host;
+    char* server_port;
+    if (!gpr_split_host_port(uri->path[0] == '/' ? uri->path + 1 : uri->path,
+                             &server_host, &server_port)) {
+      gpr_log(GPR_INFO,
+              "unable to split host and port, not checking no_proxy list for "
+              "host '%s'",
+              server_uri);
     } else {
       size_t uri_len = strlen(server_host);
-      size_t i = 0;
-      size_t marker = 0;
-      while (no_proxy_str[i] != 0) {
-        if (no_proxy_str[i] == ',' || no_proxy_str[i] == ' ') {
-          size_t no_proxy_len = i - marker;
-          if (no_proxy_len <= uri_len) {
-            if (strncmp(&no_proxy_str[marker], &server_host[uri_len - no_proxy_len], no_proxy_len) == 0) {
-              gpr_log(GPR_INFO, "not using proxy for host in no_proxy list '%s'",
-                      server_uri);
-              grpc_uri_destroy(uri);
-              gpr_free(*name_to_resolve);
-              gpr_free(server_host);
-              gpr_free(server_port);
-              *name_to_resolve = NULL;
-              return false;
-            }
-          }
-          marker = i + 1;
+      char** no_proxy_hosts;
+      size_t num_no_proxy_hosts;
+      gpr_string_split(no_proxy_str, NO_PROXY_SEPARATOR, &no_proxy_hosts,
+                       &num_no_proxy_hosts);
+      for (size_t i = 0; i < num_no_proxy_hosts; i++) {
+        char* no_proxy_entry = no_proxy_hosts[i];
+        size_t no_proxy_len = strlen(no_proxy_entry);
+        if (no_proxy_len <= uri_len &&
+            gpr_stricmp(no_proxy_entry, &server_host[uri_len - no_proxy_len]) ==
+                0) {
+          gpr_log(GPR_INFO, "not using proxy for host in no_proxy list '%s'",
+                  server_uri);
+          use_proxy = false;
```

Comment:
I think we can break out of the loop here, because there's no need to continue once we've found a match.